File: day3/part2/crossedwires.py



def visit_path(visited, path):
    pos = [0,0]
    step = 0

    for (direction, distance) in path:
        for i in range(distance):
            step += 1
            if direction is 'R':
                pos[1] += 1
            elif direction is 'L':
                pos[1] -= 1
            elif direction is 'U':
                pos[0] += 1
            elif direction is 'D':
                pos[0] -= 1
            else:
                print("Undefined direction ", direction)
                exit(1)

            if (pos[0], pos[1]) not in visited:
                visited[(pos[0], pos[1])] = step

    return visited


def manhattan_dist(pos1, pos2):
    x1, y1 = pos1
    x2, y2 = pos2
    return abs(x1-x2) + abs(y1-y2)


if __name__ == "__main__":
    # Read input
    path1 = [[list(move)[0], int(move[1:])] \
             for move in input().split(',')]

    path2 = [[list(move)[0], int(move[1:])] \
             for move in input().split(',')]

    # Mark visited positions
    # Same position can be visited twice by same wire
    # so we have two distinct lists of visited positions
    visited1 = visit_path({}, path1)
    visited2 = visit_path({}, path2)

    # Find position with fewest combined steps
    # from (0,0) that has been visited by both wires
    minsteps = 100000000
    minpos = [-1, -1]
    for pos in visited1.keys():
        if pos in visited2:
            print(pos)
            auxsteps = visited1[pos] + visited2[pos]
            if auxsteps < minsteps:
                minsteps = auxsteps
                minpos = pos

    print("----------------")
    print(minpos, minsteps)

File: day3/part2/crossedwires.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type coord struct {
	x int
	y int
}

type move struct {
	direction rune
	distance  int
}

func parseMoves(text []string) []move {
	allMoves := make([]move, len(text))
	for i, strmv := range text {
		dist, _ := strconv.Atoi(strmv[1:])
		mv := move{rune(strmv[0]), dist}
		allMoves[i] = mv
	}
	return allMoves
}

func readInput() ([]move, []move) {
	scanner := bufio.NewScanner(os.Stdin)

	scanner.Scan()
	path1 := parseMoves(strings.Split(scanner.Text(), ","))

	scanner.Scan()
	path2 := parseMoves(strings.Split(scanner.Text(), ","))

	return path1, path2
}

func visit(path []move) map[coord]int {
	visited := make(map[coord]int)
	pos := coord{0, 0}
	steps := 0

	for _, mv := range path {
		for i := 0; i < mv.distance; i++ {
			steps++

			switch mv.direction {
			case 'U':
				pos.y++
			case 'D':
				pos.y--
			case 'L':
				pos.x--
			case 'R':
				pos.x++
			}

			if !inVisited(pos, visited) {
				visited[pos] = steps
			}

		}
	}

	return visited
}

func manhattanDist(pos1, pos2 coord) int {
	return int(math.Abs(float64(pos1.x-pos2.x)) + math.Abs(float64(pos1.y-pos2.y)))
}

func inVisited(pos coord, visited map[coord]int) bool {
	if _, ok := visited[pos]; ok {
		return true
	}
	return false
}

func main() {
	path1, path2 := readInput()

	visited1 := visit(path1)
	visited2 := visit(path2)

	// Calc closest pos visited by both wires
	minpos := coord{0, 0}
	minsteps := 100000000
	for pos, auxsteps1 := range visited1 {
		if auxsteps2, ok := visited2[pos]; ok && minsteps > auxsteps1+auxsteps2 {
			minsteps = auxsteps1 + auxsteps2
			minpos = pos
		}
	}

	fmt.Println(minpos, minsteps)
}
